File: app/src/main/java/com/lostincontext/mainscreen/MainScreenViewHolder.java
package com.lostincontext.mainscreen;

import android.content.Context;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.LinearOutSlowInInterpolator;
import android.support.v7.graphics.Palette;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Interpolator;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.animation.ViewPropertyAnimation;
import com.bumptech.glide.request.target.Target;
import com.lostincontext.PlaylistLauncher;
import com.lostincontext.R;
import com.lostincontext.commons.images.palette.PaletteBitmap;
import com.lostincontext.commons.images.palette.PaletteBitmapTranscoder;
import com.lostincontext.commons.images.palette.PaletteImageViewTarget;
import com.lostincontext.commons.list.ViewHolder;
import com.lostincontext.data.playlist.Playlist;
import com.lostincontext.data.rules.FenceIconGiver;
import com.lostincontext.data.rules.Rule;
import com.lostincontext.databinding.ItemRuleBinding;

import java.util.ArrayList;
import java.util.List;

import static android.support.v4.content.ContextCompat.getColor;
import static com.lostincontext.utils.Colors.animateBackgroundColor;
import static com.lostincontext.utils.Colors.animateTextColor;

public class MainScreenViewHolder extends ViewHolder implements RequestListener<Playlist, PaletteBitmap> {

    public interface Callback {
        void onPlaylistCoverClick(Playlist playlist);
    }

    private final ItemRuleBinding binding;
    private final PaletteImageViewTarget target;
    private PaletteBitmapTranscoder transcoder;

    private final @ColorInt int defaultTextColor;
    private final @ColorInt int defaultBackgroundColor;
    private final @ColorInt int defaultIconColor;

    private final int animationDuration;
    private final Interpolator interpolator;

    ViewPropertyAnimation.Animator alphaAnimator = new ViewPropertyAnimation.Animator() {
        @Override
        public void animate(View view) {
            view.setAlpha(0f);
            view.animate()
                    .withLayer()
                    .alpha(1f)
                    .setDuration(animationDuration)
                    .setInterpolator(interpolator)
                    .start();
        }
    };

    public static MainScreenViewHolder create(LayoutInflater layoutInflater,
                                              ViewGroup parent) {
        ItemRuleBinding binding = ItemRuleBinding.inflate(layoutInflater, parent, false);
        return new MainScreenViewHolder(binding);
    }

    public MainScreenViewHolder(final ItemRuleBinding binding) {
        super(binding.getRoot());
        this.binding = binding;
        this.target = new PaletteImageViewTarget(binding.image);
        this.defaultTextColor = binding.itemTitle.getCurrentTextColor();
        Context context = binding.getRoot().getContext();
        this.defaultBackgroundColor = getColor(context,
                                               R.color.playlist_text_default_background);
        this.defaultIconColor = getColor(context,
                                         R.color.playlist_icon_default_color);

        this.transcoder = new PaletteBitmapTranscoder(context);
        this.animationDuration = context.getResources().getInteger(android.R.integer.config_longAnimTime);
        this.interpolator = new LinearOutSlowInInterpolator();
    }

    public void bindTo(Rule rule) {

        binding.setRule(rule);

        binding.setCallback(new MainScreenViewHolder.Callback() {
            @Override public void onPlaylistCoverClick(Playlist playlist) {
                PlaylistLauncher launcher = new PlaylistLauncher();
                launcher.launchPlaylist(binding.getRoot().getContext(), playlist, false);
            }
        });

        binding.textBackground.setBackgroundColor(defaultBackgroundColor);
        binding.itemTitle.setTextColor(defaultTextColor);

        Glide.with(binding.getRoot().getContext())
                .load(rule.getPlaylist())
                .asBitmap()
                .transcode(transcoder, PaletteBitmap.class)
                .animate(alphaAnimator)
                .listener(this)
                .into(target);

        List<Integer> icons = new ArrayList<>();
        rule.getFenceVM().giveIcon(new FenceIconGiver(), icons);
        if (icons != null && !icons.isEmpty()) {
            if (icons.size() >= 3) {
                binding.ic3.setImageResource(icons.get(2));
            }
            if (icons.size() >= 2) {
                binding.ic2.setImageResource(icons.get(1));
            }
            binding.ic1.setImageResource(icons.get(0));

        }
    }

    //region requestListener
    @Override
    public boolean onException(Exception e,
                               Playlist model,
                               Target<PaletteBitmap> target,
                               boolean isFirstResource) {
        return false;
    }

    @Override
    public boolean onResourceReady(PaletteBitmap resource,
                                   Playlist model,
                                   Target<PaletteBitmap> target,
                                   boolean isFromMemoryCache,
                                   boolean isFirstResource) {
        applyPalette(resource, !isFromMemoryCache);
        return false;
    }
    //endregion requestListener

    private void applyPalette(PaletteBitmap resource, boolean shouldAnimate) {
        Palette palette = resource.palette;
        if (palette == null) return;
        Palette.Swatch swatch = getSwatch(palette);
        if (swatch == null) return;
        if (shouldAnimate) {

            animateTextColor(binding.itemTitle,
                             binding.itemTitle.getCurrentTextColor(),
                             swatch.getTitleTextColor(),
                             animationDuration,
                             interpolator);

            animateBackgroundColor(binding.textBackground,
                                   defaultBackgroundColor,
                                   swatch.getRgb(),
                                   animationDuration,
                                   interpolator);

        } else {
            binding.itemTitle.setTextColor(swatch.getTitleTextColor());
            binding.textBackground.setBackgroundColor(swatch.getRgb());
        }

    }

    @Nullable
    private static Palette.Swatch getSwatch(Palette palette) {
        Palette.Swatch swatch = palette.getVibrantSwatch();
        if (swatch == null) swatch = palette.getDarkVibrantSwatch();
        if (swatch == null) swatch = palette.getLightVibrantSwatch();
        if (swatch == null) {
            List<Palette.Swatch> swatches = palette.getSwatches();
            if (swatches.size() != 0) swatch = swatches.get(0);
        }
        return swatch;
    }

    @Override public boolean onFailedToRecycleView() {
        binding.image.clearAnimation();
        binding.textBackground.clearAnimation();
        binding.itemTitle.clearAnimation();
        return true;
    }
}

File: app/src/main/java/com/lostincontext/mainscreen/MainScreenViewHolder.kt
package com.lostincontext.mainscreen

import android.support.annotation.ColorInt
import android.support.v4.content.ContextCompat.getColor
import android.support.v4.view.animation.LinearOutSlowInInterpolator
import android.support.v7.graphics.Palette
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.animation.Interpolator
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.animation.ViewPropertyAnimation
import com.bumptech.glide.request.target.Target
import com.lostincontext.R
import com.lostincontext.commons.images.palette.PaletteBitmap
import com.lostincontext.commons.images.palette.PaletteBitmapTranscoder
import com.lostincontext.commons.images.palette.PaletteImageViewTarget
import com.lostincontext.commons.list.ViewHolder
import com.lostincontext.data.playlist.Playlist
import com.lostincontext.data.rulesV2.Rule
import com.lostincontext.databinding.ItemRuleBinding
import com.lostincontext.utils.animateBackgroundColor
import com.lostincontext.utils.animateTextColor
import java.util.*

class MainScreenViewHolder(private val binding: ItemRuleBinding,
                           private val callback: Callback) : ViewHolder(binding.root),
                                                             RequestListener<Playlist?, PaletteBitmap> {

    interface Callback {
        fun onPlaylistCoverClick(playlist: Playlist)
        fun onRuleLongClick(rule: Rule) : Boolean
    }

    private val target: PaletteImageViewTarget
    private val transcoder: PaletteBitmapTranscoder

    @ColorInt private val defaultTextColor: Int
    @ColorInt private val defaultBackgroundColor: Int
    @ColorInt private val defaultIconColor: Int

    private val animationDuration: Long
    private val interpolator: Interpolator

    internal val alphaAnimator: ViewPropertyAnimation.Animator


    init {
        target = PaletteImageViewTarget(binding.image)
        defaultTextColor = binding.itemTitle.currentTextColor
        val context = binding.root.context
        defaultBackgroundColor = getColor(context,
                                          R.color.playlist_text_default_background)
        defaultIconColor = getColor(context,
                                    R.color.playlist_icon_default_color)

        transcoder = PaletteBitmapTranscoder(context)
        animationDuration = context.resources.getInteger(android.R.integer.config_longAnimTime).toLong()
        interpolator = LinearOutSlowInInterpolator()

        alphaAnimator = ViewPropertyAnimation.Animator { view ->
            view.alpha = 0f
            view.animate().withLayer()
                    .alpha(1f)
                    .setDuration(animationDuration)
                    .setInterpolator(interpolator)
                    .start()
        }
    }

    fun bindTo(rule: Rule) {

        binding.rule = rule
        binding.callback = callback

        binding.textBackground.setBackgroundColor(defaultBackgroundColor)
        binding.itemTitle.setTextColor(defaultTextColor)

        Glide.with(binding.root.context)
                .load(rule.playlist)
                .asBitmap()
                .transcode(transcoder, PaletteBitmap::class.java)
                .animate(alphaAnimator)
                .listener(this)
                .into(target)

        /* val icons = ArrayList<Int>()
        rule.fenceVM.giveIcon(FenceIconGiver(), icons)
         if (!icons.isEmpty()) {
             if (icons.size >= 3) {
                 binding.ic3.setImageResource(icons[2])
             }
             if (icons.size >= 2) {
                 binding.ic2.setImageResource(icons[1])
             }
             binding.ic1.setImageResource(icons[0])

         }*/
    }

    //region requestListener
    override fun onException(e: Exception,
                             model: Playlist?,
                             target: Target<PaletteBitmap>,
                             isFirstResource: Boolean): Boolean {
        return false
    }

    override fun onResourceReady(resource: PaletteBitmap,
                                 model: Playlist?,
                                 target: Target<PaletteBitmap>,
                                 isFromMemoryCache: Boolean,
                                 isFirstResource: Boolean): Boolean {
        applyPalette(resource, !isFromMemoryCache)
        return false
    }
    //endregion requestListener

    private fun applyPalette(resource: PaletteBitmap, shouldAnimate: Boolean) {
        val palette = resource.palette
        val swatch = getSwatch(palette) ?: return
        if (shouldAnimate) {

            binding.itemTitle.animateTextColor(binding.itemTitle.currentTextColor,
                                               swatch.titleTextColor,
                                               animationDuration,
                                               interpolator)

            binding.textBackground.animateBackgroundColor(defaultBackgroundColor,
                                                          swatch.rgb,
                                                          animationDuration,
                                                          interpolator)

        } else {
            binding.itemTitle.setTextColor(swatch.titleTextColor)
            binding.textBackground.setBackgroundColor(swatch.rgb)
        }

    }

    override fun onFailedToRecycleView(): Boolean {
        binding.image.clearAnimation()
        binding.textBackground.clearAnimation()
        binding.itemTitle.clearAnimation()
        return true
    }

    companion object {

        fun create(layoutInflater: LayoutInflater,
                   parent: ViewGroup,
                   callback: Callback): MainScreenViewHolder {
            val binding = ItemRuleBinding.inflate(layoutInflater, parent, false)
            return MainScreenViewHolder(binding, callback)
        }

        private fun getSwatch(palette: Palette): Palette.Swatch? {
            var swatch: Palette.Swatch? = palette.vibrantSwatch
            if (swatch == null) swatch = palette.darkVibrantSwatch
            if (swatch == null) swatch = palette.lightVibrantSwatch
            if (swatch == null) {
                val swatches = palette.swatches
                if (swatches.size != 0) swatch = swatches[0]
            }
            return swatch
        }
    }
}

